docs/source/extensions_tutorial/3_spec_api.rst

```diff
@@ -0,0 +1,173 @@
+:py:class:`~pynwb.spec.NWBGroupSpec` class, which provides a python API for defining
+`NWB Group <https://schema-language.readthedocs.io/en/latest/specification_language_description.html#groups>`_ objects.
+
+.. code-block:: python
+
+    from pynwb.spec import NWBGroupSpec
+
+    spec = NWBGroupSpec(
+        neurodata_type_def='MyType',
+        neurodata_type_inc='NWBDataInterface',
+        doc='A custom NWB type',
+        name='quux',
+        attributes=[...],
+        datasets=[...],
+        groups=[...],
+        links=[...]
+    )
+
+``neurodata_type_def`` and ``neurodata_type_inc`` define the neurodata type with the following rules:
+
+- ``neurodata_type_def`` declares the name of the neurodata type.
+- ``neurodata_type_inc`` indicates what data type you are extending (Groups must extend Groups, and Datasets must extend Datasets).
+- To define a new neurodata type that does not extend an existing type, use ``neurodata_type_def`` and not ``neurodata_type_inc``.
+- To use a type that has already been defined, use ``neurodata_type_inc`` and not ``neurodata_type_def``.
+- You can define a group that is not a neurodata type by omitting both ``neurodata_type_def`` and ``neurodata_type_inc``.
+
+.. tip::
+    Although you have the option not to, there are several advantages to defining new groups and neurodata types.
+    Neurodata types can be reused in multiple places in the schema, and can be linked to, while groups that are not
+    neurodata types cannot. You can also have multiple neurodata type groups of the same type in the same group,
+    whereas groups that are not neurodata types are limited to 0 or 1. Most of the time, we would recommend making a
+    group a neurodata type. It is also generally better to extend your neurodata type from an existing type. Look
+    through the `NWB schema <https://nwb-schema.readthedocs.io/en/latest/>`_ to see if a core neurodata type would
+    work as a base for your new type. If no existing type works, consider extending
+    :py:class:`~pynwb.base.NWBDataInterface``, which allows you to add the object to a processing module.
+
+``doc`` is a required argument that describes the purpose of the neurodata type.
+
+``name`` is an optional argument that indicates the name of the Group that is written to the file. If this argument
+is omitted, users will be required to enter a ``name`` field when creating instances of this neurodata type in the API.
+You also have the option of specifying ``default_name``, in which case this name will be used as the name of the group
+if no other name is provided in the PyNWB API.
+
+``attributes``, ``datasets``, ``groups``, and ``links`` are all optional arguments that take lists of the
+corresponding ``NWBSpec`` classes.
+
+``quantity`` indicates the number of instances of this group that are allowed. See options
+`here <https://schema-language.readthedocs.io/en/latest/specification_language_description.html#quantity>`_.
+
```

Comment:
I would avoid defining the arguments is this great detail here both to keep the tutorial easier to parse and follow for readers as well as to avoid overlap with the API docs. As a compromise, I would suggest to simply add short comments on the code example and link to the API docs for details for these parameters. For the neurodata_type_inc/def keys I think the discussion here makes sense because they provide important details for the tutorial but for items like ``doc`` or ``name`` should suffice.